docs/tutorials/blackjack_tutorial.py

```diff
@@ -0,0 +1,361 @@
+print('truncated:', truncated)
+print('info:', info)
+
+
+######################################################################
+# Once ``terminated = True`` or ``truncated=True``, we should stop the
+# current episode and begin a new one with ``env.reset()``. If you
+# continue executing act`ons without resetting the environment, it still
+# responds but the output won’t be useful for training (it might even be
+# harmful if the agent learns on invalid data).
+# 
+
+
+######################################################################
+# Building an agent
+# -----------------
+# 
+# Let’s build a ``Q-learning agent`` to solve *Blackjack-v1*! We’ll need
+# some functions for picking an action and updating the agents action
+# values. To ensure that the agents expores the environment, one possible
+# solution is the ``epsilon-greedy`` strategy, where we pick a random
+# action with the percentage ``epsilon`` and the greedy action (currently
+# valued as the best) ``1 - epsilon``.
+# 
+
+class BlackjackAgent():
+    
+    def __init__(self, lr=1e-3, epsilon=0.1, epsilon_decay=1e-4):
+        """
+        Initialize an Reinforcement Learning agent with an empty dictionary 
+        of state-action values (q_values), a learning rate and an epsilon.
+        """
+        self.q_values = defaultdict(lambda: np.zeros(env.action_space.n)) # maps a state to action values
+        self.lr = lr
+        self.epsilon = epsilon
+        self.epsilon_decay = epsilon_decay
+    
+    def get_action(self, state):
+        """
+        Returns the best action with probability (1 - epsilon) 
+        and a random action with probability epsilon to ensure exploration.
+        """
+        # with probability epsilon return a random action to explore the environment
+        if np.random.random() < self.epsilon:
+            action = env.action_space.sample()
+            
+        # with probability (1 - epsilon) act greedily (exploit)
+        else:
+            action = np.argmax(self.q_values[state])
+        return action
+    
+    def update(self, state, action, reward, next_state, done):
+        """
+        Updates the Q-value of an action.
+        """
+        old_q_value = self.q_values[state][action]
+        max_future_q = np.max(self.q_values[next_state])
+        target = reward + self.lr * max_future_q * (1 - done)
+        self.q_values[state][action] = (1 - self.lr) * old_q_value + self.lr * target
+    
+    def decay_epsilon(self):
+        self.epsilon = self.epsilon - epsilon_decay
+
+
+######################################################################
+# To train the agent, we will let the agent play one episode (one complete
+# game is called an episode) at a time and then update it’s Q-values after
+# each episode. The agent will have to experience a lot of episodes to
+# explore the environment sufficiently.
+# 
+# Now we should be ready to build the training loop.
+# 
+
+# hyperparameters
+epsilon = 0.6
+n_episodes = 300_000
```

Comment:
Can you optimize this value to make the script finish faster without hurting the results? We execute the tutorials when building the docs so making them fast is important.